pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go

```diff
@@ -0,0 +1,453 @@
+	}
+
+	return selectorsString, nil
+}
+
+func (s *sigstoreImpl) SetRekorURL(rekorURL string) error {
+	if rekorURL == "" {
+		return errors.New("rekor URL is empty")
+	}
+	rekorURI, err := url.Parse(rekorURL)
+	if err != nil {
+		return fmt.Errorf("failed parsing rekor URI: %w", err)
+	}
+	if rekorURI.Scheme != "" && rekorURI.Scheme != "https" && rekorURI.Scheme != "http" {
+		return fmt.Errorf("invalid rekor URL Scheme %q", rekorURI.Scheme)
+	}
+	if rekorURI.Host == "" {
+		return fmt.Errorf("host is required on rekor URL")
+	}
+	s.rekorURL = *rekorURI
+	return nil
+}
+
+func getSignatureSubject(signature oci.Signature) (string, error) {
+	if signature == nil {
+		return "", errors.New("signature is nil")
+	}
+	ss := payload.SimpleContainerImage{}
+	pl, err := signature.Payload()
+	if err != nil {
+		return "", err
+	}
+	if err := json.Unmarshal(pl, &ss); err != nil {
+		return "", err
+	}
+	cert, err := signature.Cert()
+	if err != nil {
+		return "", fmt.Errorf("failed to access signature certificate: %w", err)
+	}
+
+	if cert != nil {
+		return certSubject(cert), nil
+	}
+	if len(ss.Optional) > 0 {
+		if subjString, ok := ss.Optional["subject"]; ok {
+			if subj, ok := subjString.(string); ok {
+				return subj, nil
+			}
+		}
+	}
+
+	return "", errors.New("no subject found in signature")
+}
+
+func getBundleSignatureContent(bundle *bundle.RekorBundle) (string, error) {
+	if bundle == nil {
+		return "", errors.New("bundle is nil")
+	}
+	body64, ok := bundle.Payload.Body.(string)
+	if !ok {
+		returnedType := fmt.Sprintf("expected payload body to be a string but got %T instead", bundle.Payload.Body)
+		return "", errors.New(returnedType)
+	}
+	body, err := base64.StdEncoding.DecodeString(body64)
+	if err != nil {
+		return "", err
+	}
+	var bundleBody BundleBody
+	if err := json.Unmarshal(body, &bundleBody); err != nil {
+		return "", fmt.Errorf("failed to parse bundle body: %w", err)
+	}
+
+	if bundleBody.Spec.Signature.Content == "" {
+		return "", errors.New("bundle payload body has no signature content")
+	}
+
+	return bundleBody.Spec.Signature.Content, nil
+}
+
+func selectorsToString(selectors SelectorsFromSignatures, containerID string) []string {
+	var selectorsString []string
+	if selectors.Subject != "" {
+		selectorsString = append(selectorsString, fmt.Sprintf("%s:image-signature-subject:%s", containerID, selectors.Subject))
+	}
+	if selectors.Content != "" {
+		selectorsString = append(selectorsString, fmt.Sprintf("%s:image-signature-content:%s", containerID, selectors.Content))
+	}
+	if selectors.LogID != "" {
+		selectorsString = append(selectorsString, fmt.Sprintf("%s:image-signature-logid:%s", containerID, selectors.LogID))
+	}
+	if selectors.IntegratedTime != "" {
+		selectorsString = append(selectorsString, fmt.Sprintf("%s:image-signature-integrated-time:%s", containerID, selectors.IntegratedTime))
+	}
+	return selectorsString
+}
+
+func certSubject(c *x509.Certificate) string {
```

Comment:
FYI, we added another subject recently that could be stored in a different x.509 extension, see https://github.com/sigstore/cosign/blob/983c36449ad0b97130bf0392ffda8ce0158de4e3/pkg/signature/keys.go#L238